File: Data Structures & Algorithms/interleaving-string/submission-1.cpp
class Solution {
    bool check(string& s1, string& s2, string& s3) {
        int i=0,j=0,k=0;
        while(k<s3.size()&&(i<s1.size()||j<s2.size())) {
            int tmp = k;
            while(k<s3.size()&&i<s1.size()&&s3[k]==s1[i]) {
                k++;
                i++;
            }
            cout<<i<<'-'<<k<<'\n';
            if(tmp==k) return false;
            if(k==s3.size()) break;
            tmp=k;
            while(k<s3.size()&&j<s2.size()&&s3[k]==s2[j]) {
                k++;
                j++;
            }
            if(tmp==k) return false;
            cout<<j<<'='<<k<<'\n';
        }
        return i==s1.size()&&j==s2.size()&&k==s3.size();
    }
public:
    bool isInterleave(string s1, string s2, string s3) {
        bool ans = false;
        ans = ans || (s1[0]==s3[0]&&check(s1, s2, s3));
        ans = ans || (s2[0]==s3[0]&&check(s2, s1, s3));
        return ans;
    }
};

File: Data Structures & Algorithms/n-queens-ii/submission-1.java
class Solution {
    int[][] dir =
        new int[][] {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, -1}, {-1, 1}};

    boolean check(Set<Map.Entry<Integer, Integer>> st, int x, int y, int n, int dx, int dy) {
        int i = dx, j = dy;
        while (x+i > 0 && y+j > 0 && x + i <= n && y + j <= n) {
            if (st.contains(Map.entry(x + i, y + j)))
                return false;
            i += dx;
            j += dy;
        }
        return true;
    }

    public int calc(Set<Map.Entry<Integer, Integer>> st, int row, int n) {
        if (row == 0) {
            return 1;
        }
        int res = 0;
        for (int col = 1; col <= n; col++) {
            boolean safe = true;
            for (int[] dr : dir) {
                safe = safe && check(st, row, col, n, dr[0], dr[1]);
            }
            if (safe) {
                Map.Entry<Integer, Integer> entry = Map.entry(row, col);
                st.add(entry);
                res += calc(st, row - 1, n);
                st.remove(entry);
            }
        }
        return res;
    }

    public int totalNQueens(int n) {
        Set<Map.Entry<Integer, Integer>> st = new HashSet<>();
        return calc(st, n, n);
    }
}